shoppingcart/server/main.go

```diff
@@ -0,0 +1,151 @@
+	for k := range itemCosts {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		actionButton := fmt.Sprintf("<a href=\"/action?type=add&id=%s\">"+
+			"<button style=\"background-color:#4CAF50;\">Add to Cart</button></a>", k)
+		dollars := float64(itemCosts[k]) / 100
+		_, _ = fmt.Fprintf(w, "<tr><td>%s</td><td>$%.2f</td><td>%s</td></tr>", k, dollars, actionButton)
+	}
+	_, _ = fmt.Fprint(w, "</table><h3>Current items in cart:</h3>"+
+		"<table border=1><tr><th>Item</th><th>Quantity</th><th>Action</th>")
+
+	if len(cartState.Items) == 0 {
+		updateWithStartCart("", "")
+	}
+
+	// List current items in cart
+	keys = make([]string, 0)
+	for k := range cartState.Items {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		removeButton := fmt.Sprintf("<a href=\"/action?type=remove&id=%s\">"+
+			"<button style=\"background-color:#f44336;\">Remove Item</button></a>", k)
+		_, _ = fmt.Fprintf(w, "<tr><td>%s</td><td>%d</td><td>%s</td></tr>", k, cartState.Items[k], removeButton)
+	}
+	_, _ = fmt.Fprint(w, "</table>")
+}
+
+func actionHandler(w http.ResponseWriter, r *http.Request) {
+	actionType := r.URL.Query().Get("type")
+	switch actionType {
+	case "add", "remove", "checkout", "":
+	default:
+		log.Fatalln("Invalid action type:", actionType)
+	}
+	id := r.URL.Query().Get("id")
+
+	updateWithStartCart(actionType, id)
+
+	if actionType != "" {
+		listHandler(w, r)
+	}
+}
+
+func updateWithStartCart(actionType string, id string) {
+	ctx := context.Background()
+	startWorkflowOp := workflowClient.NewWithStartWorkflowOperation(client.StartWorkflowOptions{
+		ID:        "shopping-cart-workflow" + workflowIdNumber,
+		TaskQueue: shoppingcart.TaskQueueName,
+		// WorkflowIDConflictPolicy is required when using UpdateWithStartWorkflow.
+		// Here we use USE_EXISTING, because we want to reuse the running workflow, as it
+		// is long-running and keeping track of our cart state.
+		WorkflowIDConflictPolicy: enumspb.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING,
+	}, shoppingcart.CartWorkflow)
+
+	updateOptions := client.UpdateWorkflowOptions{
+		UpdateName:   shoppingcart.UpdateName,
+		WaitForStage: client.WorkflowUpdateStageCompleted,
+		Args:         []interface{}{actionType, id},
+	}
+	option := client.UpdateWithStartWorkflowOptions{
+		StartWorkflowOperation: startWorkflowOp,
+		UpdateOptions:          updateOptions,
+	}
+	updateHandle, err := workflowClient.UpdateWithStartWorkflow(ctx, option)
+	if err != nil {
+		// For example, a client-side validation error (e.g. missing conflict
+		// policy or invalid workflow argument types in the start operation), or
+		// a server-side failure (e.g. failed to start workflow, or exceeded
+		// limit on concurrent update per workflow execution).
+		log.Fatalln("Error issuing update-with-start:", err)
+	}
+
+	// If someone has checked out their cart, this completes the workflow.
+	// We then want to create a new workflow for the next user to shop.
```

Comment:
I'm not qute following this -- it makes it sound like the site would only handle one concurrent shopper. In my experience with e-commerce sites, a "session ID" is obtained from a cookie submitted by the end user client, and there's a database table containing session info keyed by session ID, from which an active shopping cart can be obtained. So in the Python sample I faked this by hard-coding a session ID, and using it as the workflow ID for the UwS operations. https://github.com/temporalio/samples-python/blob/main/message_passing/update_with_start/lazy_initialization/starter.py#L34